### PR TITLE
[web] remove obsolete object caches; simplify native object management

### PR DESCRIPTION
**Pull request by yjbanov**
Remove obsolete object caches and introduce a simpler way to manage native objects:

* Remove the unused `SynchronousSkiaObjectCache`.
* Introduce new library `native_memory.dart` that's smaller and simpler than `skia_object_cache.dart`.
* Introduce two types of native object references:
  * `UniqueRef` a reference with a unique Dart object owner.
  * `CountedRef` a ref-counted reference with multiple Dart object owners.
* All native references use GC (via `FinalizationRegistry`) as a back-up.
* The new library removes everything related to object resurrection that was needed only in browsers that didn't support `FinalizationRegistry`. All browsers support it now.
* Remove the ad hoc `SkParagraph` cache that predates the introduction of `Paragraph.dispose`.
* Rewrite `CkParagraph` in terms of `UniqueRef`.
* Rewrite `CkImage` in terms of `CountedRef`; delete `SkiaObjectBox`.

This PR does not migrate all objects from the old `skia_object_cache.dart` to `native_memory.dart`. That would be too big of a change. The migration can be done in multiple smaller PRs.

This also removes a few unnecessary relayouts observed in https://github.com/flutter/flutter/issues/120921, but not all of them (more details in https://github.com/flutter/flutter/issues/120921#issuecomment-1481958762)

### CODE REVIEW

File: lib/web_ui/lib/src/engine/canvaskit/native_memory.dart [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+///
+/// To prevent memory leaks, the underlying C++ object is deleted by the GC if
+/// it wasn't previously disposed of explicitly.
+class UniqueRef<T extends Object> {
+  UniqueRef(Object owner, T nativeObject, this._debugOwnerLabel) {
+    _nativeObject = nativeObject;
+    if (Instrumentation.enabled) {
+      Instrumentation.instance.incrementCounter('$_debugOwnerLabel Created');
+    }
+    _finalizationRegistry.register(owner, UniqueRef);
```
Comment (hterkelsen) [inline lib/web_ui/lib/src/engine/canvaskit/native_memory.dart:33]: should this be `finalizationRegistry.register(owner, this)`?

**Review by hterkelsen — CHANGES_REQUESTED** [review]
We should add a test for when the finalization registry disposes GC'ed objects to test the line I pointed out in my review

File: lib/web_ui/lib/src/engine/canvaskit/native_memory.dart (hunk 1 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/canvaskit/native_memory.dart]: Done. PTAL

**Comment by yjbanov** [comment]
> We should add a test for when the finalization registry disposes GC'ed objects to test the line I pointed out in my review

Done. PTAL.

**Review by hterkelsen — APPROVED** [review]
LGTM

**Pull request merged**